crates/oxc_linter/src/rules/eslint/no_useless_assignment.rs

```diff
@@ -0,0 +1,1219 @@
+);
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Operation {
+    Read = 0,
+    Write = 1,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct OpAtNode {
+    pub op: Operation,
+    pub node: NodeId,
+}
+
+pub type BlockOps = Vec<(SymbolId, Vec<OpAtNode>)>;
+
+pub type CfgOps = IndexVec<BasicBlockId, Option<BlockOps>>;
+
+pub struct TraverseState<'a> {
+    // We reuse this single bitset for both Live-Out (temp) and Live-In (final)
+    live: BitSet<'a>,
+}
+
+impl<'a> TraverseState<'a> {
+    pub fn new(num_symbols: usize, allocator: &'a Allocator) -> Self {
+        Self { live: BitSet::new_in(num_symbols, allocator) }
+    }
+}
+
+pub type CfgTraverseState<'a> = IndexVec<BasicBlockId, TraverseState<'a>>;
+
+impl Rule for NoUselessAssignment {
+    fn run_once(&self, ctx: &LintContext) {
+        let allocator = Allocator::default();
+        let num_symbols = ctx.scoping().symbols_len();
+        let num_blocks = ctx.cfg().basic_blocks.len();
+
+        let mut cfg_ops: CfgOps = IndexVec::with_capacity(num_blocks);
+        cfg_ops.resize_with(num_blocks, || None);
+
+        // Track symbols that are read at least once globally, no-useless assignment ignore variables that are never read anywhere,
+        // assume no-unused-vars will catch them
+        let mut used_symbols = BitSet::new_in(num_symbols, &allocator);
+
+        let mut cfg_traverse_state: CfgTraverseState<'_> =
+            CfgTraverseState::with_capacity(num_blocks);
+        // Pre-fill with empty BitSets
+        cfg_traverse_state.resize_with(num_blocks, || TraverseState::new(num_symbols, &allocator));
+
+        //walk through all symbols, collect their operations from declarations and references
+        for symbol_id in ctx.scoping().symbol_ids() {
+            let decl_node = ctx.symbol_declaration(symbol_id);
+
+            if let AstKind::VariableDeclarator(var_decl) = decl_node.kind() {
+                //skip const declarations
+                if let AstKind::VariableDeclaration(var_declaration) =
+                    ctx.nodes().parent_node(decl_node.id()).kind()
+                    && var_declaration.kind == VariableDeclarationKind::Const
+                {
+                    continue;
+                }
+
+                // Skip function and arrow function assignments
+                if !matches!(
+                    &var_decl.init,
+                    Some(
+                        Expression::FunctionExpression(_) | Expression::ArrowFunctionExpression(_)
+                    )
+                ) {
+                    let block_id =
+                        Self::get_basic_block_id(ctx, ctx.nodes().cfg_id(decl_node.id()));
+
+                    // Ensure outer slot exists
+                    if cfg_ops[block_id].is_none() {
+                        cfg_ops[block_id] = Some(Vec::new());
+                    }
+                    let block_ops_vec = cfg_ops[block_id].as_mut().unwrap();
```

Comment:
```suggestion
                    let block_ops_vec = cfg_ops[block_id].get_or_insert_with(Vec::new);
```

This should be slightly faster as the copiler can drop the `unwrap` call as it knows that the entry has to exist